src/databricks/labs/ucx/assessment/sequencing.py

```diff
@@ -0,0 +1,313 @@
+# but we use list for the required mutability
+QueueEntry = list[int | MigrationNode | str]
+
+
+class PriorityQueue:
+    """A migration node priority queue.
+
+    Note:
+        This implementation does not support threading safety as that is not required.
+
+    Source:
+        See https://docs.python.org/3/library/heapq.html#priority-queue-implementation-notes on the changes below
+        to handle priority changes in the task. Also, the _UPDATED marker is introduced to avoid updating removed nodes.
+    """
+
+    _REMOVED = "<removed>"  # Mark removed items
+
+    def __init__(self):
+        self._entries: list[QueueEntry] = []
+        self._entry_finder: dict[MigrationNode, QueueEntry] = {}
+        self._counter = itertools.count()  # Tiebreaker with equal priorities, then "first in, first out"
+
+    def put(self, priority: int, task: MigrationNode) -> None:
+        """Put or update task in the queue.
+
+        The lowest priority is retrieved from the queue first.
+        """
+        if task in self._entry_finder:
+            self._remove(task)
+        count = next(self._counter)
+        entry = [priority, count, task]
+        self._entry_finder[task] = entry
+        heapq.heappush(self._entries, entry)
+
+    def get(self) -> MigrationNode | None:
+        """Gets the tasks with lowest priority."""
+        while self._entries:
+            _, _, task = heapq.heappop(self._entries)
+            if task == self._REMOVED:
+                continue
+            assert isinstance(task, MigrationNode)
+            self._remove(task)
+            # Ignore type because heappop returns Any, while we know it is an QueueEntry
+            return task
+        return None
+
+    def _remove(self, task: MigrationNode) -> None:
+        """Remove a task from the queue."""
+        entry = self._entry_finder.pop(task)
+        entry[2] = self._REMOVED
+
+
+class MigrationSequencer:
+    """Sequence the migration dependencies in order to execute the migration.
+
+    Similar to the other graph logic, we first build the graph by registering dependencies, then we analyse the graph.
+    Analysing the graph in this case means: computing the migration sequence in `meth:generate_steps`.
+    """
+
+    def __init__(self, ws: WorkspaceClient, admin_locator: AdministratorLocator):
+        self._ws = ws
+        self._admin_locator = admin_locator
+        self._counter = itertools.count()
+        self._nodes: dict[MigrationNodeKey, MigrationNode] = {}
+
+        # Outgoing references contains edges in the graph pointing from a node to a set of nodes that the node
+        # references. These references follow the API references, e.g. a job contains tasks in the
+        # `jobs.Job.settings.tasks`, thus a job has an outgoing reference to each of those tasks.
+        self._outgoing_references: dict[MigrationNodeKey, set[MigrationNode]] = defaultdict(set)
+
+    def register_job(self, job: jobs.Job) -> MaybeMigrationNode:
```

Comment:
@nfx : Please comment on the signature. Do you think this is the right one? An other option would be to add an external facing `register_jobs()` that loops over all jobs it can find and then calls this method (which we can make protected). 

Or we could rename this to a `JobMigrationSequencer` with a `register` and `generate_steps` public method. Later we can add a `DashboardMigrationSequencer` and abstract away a parent `Sequencer` class enforcing this API

@ericvergnaud : You added the dependency graph here. Why did you do that? It was not used (yet)

Reply:
Thought about it a bit more, maybe we need to make this method protected access as well and introduce a `build_dependency_graph` instead. I am thinking that it makes sense to get this closer to the other linter code.

Open questions I have:
- Should this method expect a job or job_id. The later is similar to the cluster: get the job given the id, if not found return problem, otherwise parse the job like the code below.
- Should we distinguish between "resolve" and "register", where we do both in the same method now. If so, should we do an early return of the resolve problems, if any.
- Should the `job_migration_node` get the problems from the `tasks` and `clusters`? I think it makes more sense to isolate the problems to the node it **directly** applies too and surface all the problems when building the dependency graph.

Reply:
`register_jobs()` is preferred.

> Should this method expect a job or job_id. The later is similar to the cluster: get the job given the id, if not found return problem, otherwise parse the job like the code below.

`jobs.Job` - we get tasks from it.

> Should we distinguish between "resolve" and "register", where we do both in the same method now. If so, should we do an early return of the resolve problems, if any.

imho, we need just `register()`

> Should the job_migration_node get the problems from the tasks and clusters? I think it makes more sense to isolate the problems to the node it directly applies too and surface all the problems when building the dependency graph.

tasks and job cluster problems have to rollup to a job. less objects for users to think of. though, if you think that having separate nodes for tasks in the graph is easier to implement - go ahead with `("TASK", f"{job.job_id}/{task.task_key}")`